src/WebJobs.Extensions.DurableTask/DurableHttpRequest.cs

```diff
@@ -170,11 +174,60 @@ public override void WriteJson(JsonWriter writer, object value, JsonSerializer s
                 else
                 {
                     // Don't know how to serialize this - use default behavior, forcing TypeNameHandling.Objects to correctly serialize ITokenSource
-                    var currTypeNameSetting = serializer.TypeNameHandling;
-                    serializer.TypeNameHandling = TypeNameHandling.Objects;
-                    serializer.Serialize(writer, value);
-                    serializer.TypeNameHandling = currTypeNameSetting;
+                    var safeTokenSerializer = GetTokenSourceSerializer(serializer);
+                    safeTokenSerializer.Serialize(writer, value);
+                }
+            }
+
+            private static JsonSerializer GetTokenSourceSerializer(JsonSerializer serializer)
+            {
+                if (tokenSerializer == null)
+                {
+                    if (serializer.TypeNameHandling == TypeNameHandling.Objects
+                    || serializer.TypeNameHandling == TypeNameHandling.All)
+                    {
+                        tokenSerializer = serializer;
```

Comment:
make sure to return here, otherwise you will still be overwriting `tokenSerializer`.